### ISSUE TITLE
Provide a "focus-follow-mouse" setting

### ISSUE THREAD

**Issue by guillaumeportails** [microsoft/vscode#93772]
I am used for years (X11 times) with focus following the mouse, instead of having to click to type (which is slow, and wears).

**Comment by Souravgoswami** [comment]
This feature will be very handy particularly for Linux users like me who has the habit of enabling the focus follows mouse feature in the window manager...

**Comment by chapmanjacobd** [comment]
This is not only a Linux issue but a Windows one as well

https://superuser.com/a/1209478/76968

**Comment by benlk** [comment]
It's also a Mac issue; Terminal.app has this as a configuration via the `defaults` interface: https://osxdaily.com/2011/04/18/terminal-focus-follow-mouse-mac/

**Comment by seritools** [comment]
Would love to see this!

Windows supports it (although it can only enabled via WinAPI or registry changes), Windows Terminal also supports it since 1.7. Now only VSCode needs it for switching focus between terminal(s), open editors, and the sidebar/toolbar elements.

**Comment by RayKoopa** [comment]
Would really like to see whatever pane being under the mouse pointer getting the focus. Got used to it a lot in Blender.

**Comment by danky78** [comment]
I would love to have this. I use it for all my terminals and I always forget to click the integrated terminal in vscode, its very annoying.
I know this must be a hard to implement feature.
Thanks code team for the good work you are doing! :)

**Comment by hicklemon** [comment]
Clicking files in a multi-pane layout can lead to inadvertently moving selected text, or inadvertently canceling a text selection.

This feature would help correct these potential problems, and it would help VSCode fall better in line with Linux environments where focus-follows-cursor is heavily in use (as with https://github.com/i3/i3 and https://github.com/software-jessies-org/jessies/tree/master/terminator)

**Comment by rossim2i2** [comment]
I have a super ultra wide monitor.  with the terminal panel on the right side of the screen, this function would basically allow me to live in vscode.

**Comment by mwaeckerlin** [comment]
Yes, focus should strictly follow mouse between all sub windows! Not only between editor and terminal, but also between different editors and terminals and side bar.

**Comment by asemic-horizon** [comment]
+1 This is also an important usability feature for people on the move using trackpads who need to minimize their clicking due to RSI.

**Comment by samkshah** [comment]
This has been open for quite a while now. Anyone knows if it'll be worked on? Other duplicates have been closed, not this one 🤔

**Comment by cu** [comment]
I hope that it stays open as it is a feature I would really like to see.

**Comment by danky78** [comment]
I would love it too. Im using external terminal because this.

**Comment by citizenrich** [comment]
> hold `Ctrl` and enable focus in sections by mouse moves.
> release `Ctrl` to back focus to the previous state.

This doesn't work across the window, to go to the Terminal, for example.

**Comment by byehack** [comment]
> This doesn't work across the window, to go to the Terminal, for example.

We can ignore `release ctrl to back to previous state`.
btw I think hold a key and move mouse to active that state is the best solution.

**Comment by hrafnkellbrimar** [comment]
> hold `Ctrl` and enable focus in sections by mouse moves. _~release `Ctrl` to back focus to the previous state.~_

I think focus-follows-mouse functionality should be selectable as a persistent setting, having to press or hold down keys is not the same.

**Comment by KeithHanlan** [comment]
This is almost a safety issue. The number of times that I've started typing with my mouse in one window and focus in another is countless. One of these days, I'm going to hit the Del key while I think that I'm in an editor or terminal window and I'm going to nuke a file in the File Explorer where the Del key means "Delete permanently".

**Comment by tjmcode** [comment]
I have become so accustomed to 'focus-follows-mouse' in all other environments that it took me a few days to figure out why I was having issues with the Integrated Terminal!

**Comment by clecap** [comment]
For me the current solution unfortunately can only be *not* to use terminal windows in vscode but to use an external terminal. I am just done with destroying editor contents after moving the mouse into the terminal subwindow and forgetting to click.

**Comment by mwerle** [comment]
I just moved to a new company which largely uses VS Code as their development platform on Linux. So far, overall it's not bad, but the fact that focus doesn't follow mouse is a constant source of aggravation for me. 
I'm always typing terminal commands into my source-code or, worse (when working on a shell script) typing shell commands directly into the terminal.. it's only a matter of time before I `rm -RF *` in an important directory.

Please default Focus to OS settings and provide an override within VS Code for broken OS's or for people who want one behaviour on their Desktop and another within VS Code. Ie, options should be "Default Focus", "Focus follows Mouse", or "Click to Focus".

But seriously, this is a no-brainer to want to support, especially given the lack of hot-keys to switch between different panels (OT, but Ctrl-Alt-<direction> and some visible indicator as to what the active panel is would be nice..) .

**Comment by josephtingiris** [comment]
I would like to see this feature, too.

**Comment by unixwzrd** [comment]
> it's only a matter of time before I `rm -RF *` in an important directory.
^
+ - This... @mwerle said it.

I use the Vi/Vim editor extensions and vi mode on the command line in my shell.  It's very efficient, same keys for editing text and my command history. The edit commands are muscle memory.

Outside of VSCode all my windows on macOS have "raise on focus" and "focus follows mouse".  The raise isn't so important, but focus following the mouse from pane to pane would be great.

Because I use Vi mode on the command and in the edit panes, I can't count how many times I've moved the mouse from the editor into the terminal and begin typing and do potential damage to a file I was editing. What's worse is I may have 4 vertical panes open and go to enter commands on the command line and only after I've typed a bit notice nothing has echoed. Only to find I've just performed unknown edits to one of my edit panes and am not sure which one was active at the time. VSCode actually isn't too bad, I actually like it, but yeah, I guess I gotta change my workflow.

Please fix this before my finger suffers RSI, or I throw the keyboard and trackpad at my monitor. This can't be that difficult.

**Comment by emogenet** [comment]
Please add this. The behavior change between all the other windows on my desktop and the windows within VSCode basically make VSCode unusable for me.

Also, clicking on windows to be able to interact with them is sloooooooooooooooow, please give us that productivity boost.

**Comment by GentleHumour** [comment]
I just typed Ctrl-Q to quit out of Micro (don't judge me!) running in the integrated terminal (yes, I'm aware I have another text editor right there, but muscle memory) and accidentally quit VSCode because I forgot to click on the terminal to focus it. Needless to say, my productivity was impacted.

I would love it if the VSCode peeps could implement this feature.

**Comment by oricou** [comment]
I tried VS code few years ago, I gave up because I am too much used to have this option. I retried now, VS has plenty wonderful options, but I cannot work without focus follow mouse. So if you can add it, it would be very nice.

**Comment by hseritt** [comment]
Add it. Jesus, I could implement it over a weekend.

**Comment by BBaoVanC** [comment]
> Add it.
> 
> Jesus, I could implement it over a weekend. Don't make me do it and embarrass you.

Why don't you do it then? That's the point of open source software. Nobody here is working for your hire.

**Comment by hseritt** [comment]
Having some fun with ya's. Just go ahead and do it already.

**Comment by Jeff-Bloom-US** [comment]
Is there no persistent configuration setting for VS Code to enable the window focus to be "follows mouse"?   Just learning VS Code and can't find anything.  Found this thread that's over 3 years old.  Apparently it's never been added/fixed.   Not sure what the comment a ways back in time that says "moved to backlog" means.  Is this getting fixed sometime?

**Comment by GentleHumour** [comment]
> Is there no persistent configuration setting for VS Code to enable the window focus to be "follows mouse"?

Not yet.


> Not sure what the comment a ways back in time that says "moved to backlog" means. Is this getting fixed sometime?

At least 20 people upvoted this issue in the [Backlog Candidates](https://github.com/microsoft/vscode/milestone/107) to prevent it from being closed. It subsequently made it into the [Backlog](https://github.com/microsoft/vscode/milestone/8).

Aside: it's surprising to me that there is no search box for milestones. The backlog has about 4,000 issues open and 9,000 issues closed and it seems like the way you search this is by scrolling to the bottom for _minutes at a time_ while they lazily load and then Ctrl-F. I'm still searching, TBH.

One of the other commenters suggested the current issue might be a weekend job. Indeed, how hard *could it be*? I was curious.

```
$ cd vscode/src
$ cloc .
    4343 text files.
    4310 unique files.                                          
     184 files ignored.

3 errors:
Line count, exceeded timeout:  ./vs/workbench/api/test/node/extHostSearch.test.ts
Line count, exceeded timeout:  ./vs/workbench/contrib/extensions/test/electron-sandbox/extensionRecommendationsService.test.ts
Line count, exceeded timeout:  ./vs/workbench/contrib/interactive/browser/interactive.contribution.ts

github.com/AlDanial/cloc v 1.90  T=9.18 s (453.9 files/s, 132103.2 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
TypeScript                    3721         166480         108335         864608
JavaScript                      50           1638           2544          31171
CSS                            222           4165           1503          18846
JSON                           114              1              0           6249
HTML                            16            423             34           2968
SVG                             28             24              9           1873
Bourne Shell                     4             54             65            295
zsh                              4             33             38            161
PowerShell                       1             19             45            131
Markdown                         3             34              0            125
Fish Shell                       1             21             49            121
LESS                             1              1              1              0
-------------------------------------------------------------------------------
SUM:                          4165         172893         112623         926548
-------------------------------------------------------------------------------
$ find . -name '*.ts' -exec grep -i focus {} /dev/null \; | grep -v test | wc -l
9514
```

She's a big 'un. And one of [a couple](https://github.com/microsoft/terminal/issues/5258) of Microsoft [projects](https://superuser.com/questions/954021/how-do-you-enable-focus-follows-mouse-in-windows-10) where sloppy focus mode is a neglected feature.

Looking at those other issues reminds me: we need to specify that *raising a focused window* is a separate matter from focusing the window on hover. People who exclusively use Windows may be unaware of this. There are situations where you really *don't* want to raise that newly-focused VSCode window over the top of the application you're debugging, or whatever. Avoiding a flurry of window Z/layer management by clicking on stuff is one of the things we would like to avoid.

**Comment by BBaoVanC** [comment]
> There are situations where you really _don't_ want to raise that newly-focused VSCode window over the top of the application you're debugging, or whatever. Avoiding a flurry of window Z/layer management by clicking on stuff is one of the things we would like to avoid.

I wasn't even aware that VSCode has any floating windows/dialogs that could cause this issue, unless you're talking about the VSCode window itself at the OS level? I understand this feature request as referring to individual panes inside the VSCode window (anything else would be the OS/desktop environment's job).

**Comment by Hate9** [comment]
> > There are situations where you really _don't_ want to raise that newly-focused VSCode window over the top of the application you're debugging, or whatever. Avoiding a flurry of window Z/layer management by clicking on stuff is one of the things we would like to avoid.
> 
> I wasn't even aware that VSCode has any floating windows/dialogs that could cause this issue, unless you're talking about the VSCode window itself at the OS level? I understand this feature request as referring to individual panes inside the VSCode window (anything else would be the OS/desktop environment's job).

When you hover over a window at an OS level, and your desktop focuses-on-hover, the window is not raised, only focused, This means that you can enter text into it without raising the window. Without focus-on-hover for vscode, a user can't use this benefit for the terminal unless they had already clicked into the terminal before focusing a different window.

**Comment by unixwzrd** [comment]
> > > There are situations where you really _don't_ want to raise that newly-focused VSCode window over the top of the application you're debugging, or whatever. Avoiding a flurry of window Z/layer management by clicking on stuff is one of the things we would like to avoid.
>>
That's why they are options, you don't want it, don't turn it on. I have ny workflow, you have yours, other people have theirs, whatever works for you. Options, people want options, I used to have virtual desktops in a 3x3 arrangement and would roll off the right, left, top, down and be on the adjacent desktop simply by pushing through the edge and wrapping. It worked for me, did everyone else like it, not really, but I was never more than 1 desktop away from any of the 9 I had set up.

> > I wasn't even aware that VSCode has any floating windows/dialogs that could cause this issue, unless you're talking about the VSCode window itself at the OS level? I understand this feature request as referring to individual panes inside the VSCode window (anything else would be the OS/desktop environment's job).

I did not know this either, will have to try it out, live and learn.

> When you hover over a window at an OS level, and your desktop focuses-on-hover, the window is not raised, only focused, This means that you can enter text into it without raising the window. Without focus-on-hover for vscode, a user can't use this benefit for the terminal unless they had already clicked into the terminal before focusing a different window.

Focus-on-hover, focus-follows-mouse, raise-on-focus, raise-on-hover, then delay times for all to happen.  If you don't like the options I have, I'm not going to force you to use them, my time delays and mouse-acceleration are probably not to your liking either, but you have a different workflow, everyone has their own workflow and the tool shouldn't slow me down.

Wanna know another thing the tool shouldn't do? Shoot me in the foot on a regular basis. I use "vi mode" in edit panes, when I'm outside VSCode, everything works the way I like it.  I move the mouse over a window and it gets focus. I can't count the number of times I've moved my mouse from elsewhere on my desktop and last time I was in VSCode an edit panel was active with focus and I move back into VSCode, sliding the mouse over a terminal and start typing only to start wondering where my characters are going because latency isn't *THAT* bad, to noticing that one of my edit panels is active and has focus. Then trying to figure out what keys I hit or what mode the editor was in when I started typing... this happens on a regular basis. To compound that, there are two undo/redo buffers in the editor, one for "vi mode" and the other for undo/redo the OS provides, yeah that sucks too because they don't sync stacks.

I really wish they would fix this issue, maybe before I retire?

**Comment by Hate9** [comment]
@unixwzrd Why does your reply (specifically your reply to me) read so defensive? I wasn't even responding to you, I was explaining to *someone else* why one might want the option.

Definitely on the same page with you though, I can't *believe* this isn't an option yet.

**Comment by unixwzrd** [comment]
@Hate9 wasn't directed at you directly, the "you" is the collective you. Just a bit frustrated with the fact that this "problem" has been here for more than 4 years. It really is about individual workflow though isn't it?

Be well...

**Comment by rmcnew** [comment]
I use "focus follows mouse" on both Windows and Linux.  It is nice because I don't have to click, just move the mouse and start typing.  It is very nice and easy-to-use.

Sadly, VS Code does not have a setting to support it.  Ideally, it should focus whatever editor pane (from vertical or horizontal splits) or embedded terminal the mouse is over.

The workaround is to just create multiple VS Code windows for each file I am editing and use a separate terminal.  It feels clunky and undermines the whole point of having VS Code support multiple editor panes and an embedded terminal.

**Comment by zhuoqun-chen** [comment]
> Would really like to see whatever pane being under the mouse pointer getting the focus. Got used to it a lot in Blender.

May I ask how to set it up like this in blender?

**Comment by unixwzrd** [comment]
> I use "focus follows mouse" on both Windows and Linux. It is nice because I don't have to click, just move the mouse and start typing. It is very nice and easy-to-use.
>
> Sadly, VS Code does not have a setting to support it. Ideally, it should focus whatever editor pane (from vertical or horizontal splits) or embedded terminal the mouse is over.
> 
> The workaround is to just create multiple VS Code windows for each file I am editing and use a separate terminal. It feels clunky and undermines the whole point of having VS Code support multiple editor panes and an embedded terminal.

These are all fantastic points. Rhetorical question how does "mouse hover" work, and how is it so many extensions use it even if the window doesn't "officially" have focus?

I use it on macOS and have used focus-follows-mouse and raise-on-focus for many years, in Unix using X, Linux and X, Windows using some registry hack I believe, and macOS. This is embedded deeply in my cerebellum and is reflex from over 35 years of doing this. Not having this has caused me to hose things up quite a few times. This is the only application I have to move to the pane and click the mouse/trackpad/pointing device, and the application I spend the most time in it. It could really be a great app if some of the rough edges could be worked out, but it's the best that's available, so it's not like I'll find something else which can do as many things as VS Code does for me.

I know the panes are getting mouse over messages because I have seen them in BetterTouch Tool. Not only that there are so many extensions that use "mouse hover".  I cannot fathom having the number of mouse-hover extensions and that giving the pane the mouse is in focus absolutely boggles my mind. This cannot be that difficult to implement, and if I didn't have other things to do, I'd do it myself, but that would mean learning the code base, JS and other things I have no time for.

Now that there is integrated AI like Codeium, Copilot and others, I cycle from edit pane, to AI chat, to terminal, to AI chat, to edit pane.  Then there's selecting a file or something in the side bar, especially opening a new file.  I select the file and it opens in the last active edit pane. Great, now I've clicked in the file explorer and then rather than being able to move directly and edit the file, I have to move to the pane where the file is open again and click there. Add the AI pane which I have attached to the right side of my VS Code window and I am clicking all over the place.

I tried using iTerm2 as a workaround for the integrated terminal problems I have at times, but I cannot get iTerm2 fully integrated, even though I've read plenty of places that it's simple to do.  For some  reason it refuses to work.  Having iTerm2 would be great as that would allow me to never have to enter the integrated terminal and then my edit sessions would get focus when I move back into VS Code from iTerm2 and vice-versa.  Problem is that iTerm2 does not work with the debugger.

The macOS ChatGPT app is in similar situation as how I use iTerm2, it sits off to the right side and if I want to use it, I move past Copilot or Codeium and it auto-raises and gets focus, which is outstanding, but I have to copy and paste to it, but at least no clicking.  Move back to VS Code and it auto raises, but then I have to make sure I'm in the active pane or click in the pane to make it active, and no integration means copy and paste.

Apart from YouTube, having to go through my email, and playing with Stable Diffusion, this is the biggest source of frustration and waste of my time having to deal with this, it has cause me to make inadvertent edits to code which I have to figure out what I did and undo it once I notice my keystrokes are not going where I intended. All this creates the biggest source of frustration in my day, every day, multiple times a day.

If someone too a bit of time and actually implemented this, I have no doubt we would create a national holiday for you. Besides, it's my birthday this month, and I would love to get a really nice birthday present of focus follows mouse, so for FFS, FFM, someone please.

Sorry for the length, but this is my semi-annual post to this issue. My first was about two years ago, sorry for the length, but I feel better and I think I express a lot of the same frustration others feel with this.

Be well...

**Comment by pruneau628** [comment]
++1

**Comment by kiviuk** [comment]
It really sucks, we had focus-follows-mouse back in 1993.  32 years ago.

**Comment by unixwzrd** [comment]
> It really sucks, we had focus-follows-mouse back in 1993. 32 years ago. And Microsoft...?

X11 had it with various window managers **WAY** before then.

I looked into whether this could be done as an extension, and unfortunately no because extensions cannot be done as extensions are sandboxed and would need access to all system events.

Asking o1 about it and ddi a bit of researching myself, got this response,

>There’s no supported way for an extension to intercept mouse enter or leave events for existing VS Code panels directly. The extension API isn’t designed to expose low-level UI event hooks for the editor’s own interface elements, including the built-in sidebar, panel, and activity bar.
>
>Extensions can handle mouse events within their own contributed content—most notably via a webview—but not in the surrounding UI. Without a webview, the underlying Electron-based shell isn’t accessible to your code in any official or reliable manner.
>
>If this functionality is important, consider filing a feature request with the VS Code team or exploring ways to achieve the desired effect through custom panels (webviews) rather than existing ones. The built-in panels simply don’t provide programmatic mouse event subscriptions through the current extension API.

I put it in as a suggestion to the Codeium team for Windsurf, haven't done so with Cursor yet, but will soon.

If you want to upvote this as a feature for Windsurf, here's the link:

https://codeium.canny.io/feature-requests/p/windsurf-focus-follows-mouse-as-a-configuration-option

I'll file another lmk if I can find suggestions for Cursor.

**Comment by snwfdhmp** [comment]
Need this !

**Comment by unixwzrd** [comment]
Just my usual chime-in rant when this ends up in my email as this frustrates me every day multiple times a day...

I just sifted through the history of this and the timeframe where this has been on the radar as an issue goes way back to more than five years in the OP. This feature request goes back as far as April 2017, that's 8 - ***EIGHT*** years.  It also seems that someone at one point had a PR to handle this as well.  Here's what I managed to dig up...

- Apr 29, 2017 [Focus follows mouse #25685](https://github.com/microsoft/vscode/issues/25685)
- Feb 22, 2018 [Add a setting to activate workbench parts on mouse over #44214](https://github.com/microsoft/vscode/issues/44214)
- Mar 15, 2018 [Support sloppy focus #45926](https://github.com/microsoft/vscode/issues/45926)
- Jul 10, 2018 [Added contrib extension to have focus follow mouse (fixes #25685). #53963](https://github.com/microsoft/vscode/pull/53963)
- **This "*Issue/Discussion*" Mar 30, 2020** [Provide a "focus-follow-mouse" setting #93772](https://github.com/microsoft/vscode/issues/93772)
- May 24, 2020 [Focus on Terminal View on mouse hover #98475](https://github.com/microsoft/vscode/issues/98475)

Seems by reading the history, this was once fixed, though I didn't dig deeper into the PR. But really, come on guys, a lot of people want this and moving between terminal, AI assistants, the Explorer side-bar, various extensions side-bars, and edit panels would be so much easier. AI assistants especially.

I have ChatGPT on my desktop, which can now look into VS Code, but can't take action, and I have iTerm2 which it can also see, but not take action. Those windows use Focus-Follows-Mouse and Raise-On-Focus using a nifty little app called [AutoRaise](https://github.com/sbmpost/AutoRaise) by @sbmpost - shoutout to those guys and give'm a star... The only problem is there is not complete integration as there is with AI Agents like Copilot, Windsurf, Cursor, and others. Seems like there are so many people out there working with VSCode forks that someone might be able to put this seemingly easy to implement feature in before another eight years passes. The integration of all these apps will probably happen before this is implemented at this rate.

I promise there will be w world holiday and celebrating in the streets at this point named for the person who takes care of this, the police might have to be called and bring out their riot gear. I know my middle finger will have a lot less stress on it.

Unbelievable... FFM FFS

**Comment by jbiosca78** [comment]
I'm, so tired of this that I'm not using vscode anymore.

**Comment by ahernsean** [comment]
I just created a [new issue](https://github.com/microsoft/vscode/issues/240238) on the "Backlog Candidates" list, in hopes of refreshing this request in the eyes of the VSCode maintainers.

**Comment by tjmcode** [comment]
Thanks!

This is truly a POA, especially when moving between multiple Dev environments all day.

Timothy McGuire
Cell: 810.459.7508

**Comment by unixwzrd** [comment]
Everyone should go and upvote this one right now!

[Provide a "focus-follow-mouse" setting #240238](https://github.com/microsoft/vscode/issues/240238)

Thanks to @ahernsean for finding the right place to post it again! It even looks like someone is assigned to it now as well!

Be well...

**Comment by snwfdhmp** [comment]
YOU WANT THIS FEATURE ?? Upvote this link: https://github.com/microsoft/vscode/issues/240238

Please share !

**Comment by unixwzrd** [comment]
@snwfdhmp That "Issue" has been closed and this "issue" has finally been assigned.  People should leave that one alone since it is closed and merged with this one. This one also has someone assigned to it now *FINALLY*.  But that shouldn't mean anyone who wants this shouldn't continue to watch this entry and make noise.  Remember "The squeaky wheel gets the grease." Spread the word that this exists and get others to upvote it. The other one got enough votes in 3-4 hours to get this one brought back to life. See my comment above with [the history of this issue.] (https://github.com/microsoft/vscode/issues/93772#issuecomment-2644525035)

This has now been set as an "Accessibility" issue, hopefully we will all see some progress.

Be well...

**Comment by unixwzrd** [comment]
Oh yeah, the next thing this one needs is a "Due Date" assigned to it.

**Comment by dirkson** [comment]
It's been a couple months without movement on this issue. Could we get a due date for it?

**Comment by ahernsean** [comment]
@meganrogge, can we get a status update from you on this feature? I know people are eager to give it a try. Even if the status is "It's on the backlog, but we have no due date", it would be useful to hear.

**Comment by meganrogge** [comment]
We currently do not have plans to implement this feature. While we understand the motivation behind it, introducing this level of complexity could create a significant surface area for bugs and instability, which we do not have the bandwidth or resources to support effectively. We want to prioritize maintaining a stable and reliable experience for all users.

**Comment by mattheworris** [comment]
Has there been any work done to support focus-follows-mouse that could possibly be a start for open source contributions? A feature or development branch?

**Comment by dirkson** [comment]
It's incredibly frustrating to see such a critical usability issue treated so strangely. It's been ignored through a half dozen issues, added to the backlog, self-assigned, left open, and now soft-won'tfixed by the person who self assigned it.

I get that there's probably some internal discussion happening that we're not privy to, but... Maybe discussion for open source software should occur in the open? Just a thought.

This feature is simply the logical extension of the focus model that VSCode _already_ uses. Try moving the mouse cursor around the vscode window: Make it hover over the terminal; Make it hover over the file explorer. UI elements change to indicate that the area the mouse is hovered over is the user's current focus - The thing that they are about to interact with.

Until they go to press a keyboard button, and something in a different, un-indicated section of the UI changes. There's no other way to say it: That's bad UI design. This change would introduce a level of "complexity" that's _simpler_ than the current dual focus model.

I've been daily driving windows for a couple years now, and this still trips me up in the OS a dozen times a day. However, it's far worse inside VSCode, because VSCode gives such strong indicators of focus. And I can't imagine what linux daily drivers are experiencing, since many of them simply won't encounter software that mandates focus-follows-click outside of VSCode.

Yes, each new feature is a surface for bugs. If VSCode were a small, closed source project with limited funds, I could see that being a major factor. But it's not. I've got an option for "Zen Mode" in the UI that changes the entire UI up. Clearly there are enough resources to implement and maintain large, radical UI options, and we're not asking for a large _or_ radical UI option.

And on top of all of that - I believe there was already a PR for this very focus model. Your community is clamoring to do the work.

Focus follows mouse is small, logical, already partially-implemented UI change that would dramatically improve accessibility for many users. Please reconsider implementing it, or at least express openness to an updated PR to implement it.

Now I'll talk to you all later. I've got some clouds I've got to go shout at.

**Comment by unixwzrd** [comment]
@dirkson  heh, you rattled my cage, but I was wondering this too.  I've been too busy working on all my other stuff to look into the mechanisms of this as well. It's super annoying and in my search fro other PR's someone did mention that someone they worked with came up with a fix, but their company didn't want to have their people use unofficial "patches". I suggested that they submit their modification in a PR, because if it works, then why not get it integrated with the code-base.  *crickets*

I was also wondering when this would be seen to and a due date assigned, as it continues to trip me up daily. I mean plenty of things respond to hover messaged, so I really, and I'm trusting the devs here, but I can't understand why everyone says it's so difficult to add this feature. I have Focus-Follows-Mouse and Raise-On-Focus on macOS as an add-on and it works great. When I was forced to use Windows as part of a job I was working on, they have this feature available in the configuration.

I realize that VSCode is just one big window and the panels are inside that window, but when you move the mouse cursor from panel to panel the scroll bars come alive and are active. Try it, see?  This is also why I cannot understand that if they are intercepting the mouse inside the panel to activate the scroll bars, why not go the extra step and in addition to activating the scroll give the panel focus?

It seems to me the code is practically already there and the code which activates the scroll bars should be able to set focus as well. Seriously, unless I'm missing something earth shaking, it really seems like the code is almost already there in the existing code-base.

What am I missing that makes it so incredibly difficult that it takes almost a decade to implement?

**Comment by rmcnew** [comment]
"We want to prioritize maintaining a stable and reliable experience for all users."

You mean, "all able users".  Maybe it's easy for you to click the mouse, but it is not easy for everyone.

**Comment by unixwzrd** [comment]
This has been assigned, reassigned, bounced around-most recently about a month ago. Others have pointed out that a working solution has existed at various points, but it's never been merged. At one point, it even had a proposed fix and a date. And yet... here we are still.

With the introduction of AI agents and chat panels, combined with the explorer and terminal panes, the absence of focus-follows-mouse has become *actively harmful* to productivity. I've lost count of how many times I've selected a file in the explorer, moved to the editor panel, and hit **⌘+A**-only to select all the files in the sidebar instead, or worse.

It's the same story with typing in the chat and intending to switch to the terminal-only to accidentally send a command to the AI panel. Other apps (Xcode, OpenAI's macOS app, even old-school X11 setups from the 80s) get this right. Why not VS Code?

This isn't just annoying-it's a usability regression. As someone who's been using X11 since 1988, my muscle memory expects panels to *do what every other app does*. The fact that scrollbars activate properly when hovering over different panels shows the plumbing is already there.

There's been Reddit chatter about this-Cursor, Windsurf, and other forks are chasing it too. Accessibility isn't just screen readers; it's about minimizing friction for people whose habits were formed in real computing environments over decades. This is a quality-of-life issue and a potential source of coding errors inside the IDE.

Just make it a config option. Ship a first cut. Let the community refine it if needed. Even with AI assistants, if someone pointed me to the right file, I could probably hack a working version together-and I'm not even a JS/TS dev.

I know some earlier comments here were deleted-my original one is now nearly two years old. I remember one that said something like, *"I hope this lands before I'm on my deathbed."* That got purged.

See this entry for the complete timeline or PR's for this: [This PR history timeline](https://github.com/microsoft/vscode/issues/93772#issuecomment-2644525035)

But seriously, FFS, I'm not getting any younger, so please: FFM.

**Comment by xfinityandbeyond** [comment]
While I don't disagree with you, I think there is a more polite way to approach it:

Just because you can think of a reason why a future shouldn't exist doesn't mean that it shouldn't exist. The opposite can be said, too - just because a feature exists doesn't make a good.

And there are a slew of features in all Ides that are questionable.

That said, to your point, this should simply be an option.

**Comment by clecap** [comment]
I fully support the request for this UI feature.

My personal favorite is that Microsoft has a reason for not wanting this feature. Let's remember what happens when you switch Windows and Mac Keyboards. On Windows you are used to having the @ on AltGr-Q whereas on MacOs the identical keyboard behavior is a command-Q, which closes the application. Ooops. At least on quite a number of keyboard layouts. So every user switching from Windows to Mac Keyboards will unintentionally close their applications once in a while. Got the message, dear user?

I have been following this thread for some time and the only explanation I can come up with is this con$piracy in some Micro$oft product development center where they do not *want* to keep the UI flexible or adaptable for the user ... exactly on places where it *does* hurt. For rea$on$ of dominating the market.

**Comment by unixwzrd** [comment]
> I fully support the request for this UI feature.

Thanks, we need to keep interest up and make some noise a bit, the squeaky wheel gets the grease, as they say. This accessibility feature request continues to age.

> My personal favorite is that Microsoft has a reason for not wanting this feature. Let's remember what happens when you switch Windows and Mac Keyboards. On Windows you are used to having the @ on AltGr-Q whereas on MacOs the identical keyboard behavior is a command-Q, which closes the application. Ooops. At least on quite a number of keyboard layouts. So every user switching from Windows to Mac Keyboards will unintentionally close their applications once in a while. Got the message, dear user?

Have you tried Better Touch Tool?  It has the ability to distinguish between keyboards, and might do the trick for handling this situation for you.  I've found it handy for things like mapping my "Enter" from the "return" key to Command+Return for the ChatGPT desktop app.  Tried to do the same thing for the AI chat boxes in Windsurf and Cursor, but have haven't had any success, so I'm stuck with Shift+Return for now to add extra line break sin my prompts. It's a handy extension for macOS, I had to copy and paste a bunch of text and "paste and Match Style", so I created a shortcut for that making it much easier when pasting into the Notes.app, set it up so Command+Left-Click would do it for me.

I've accidentally hit Command+Q because the Q is placed very close to the A, S, and W keys, and if I make a typo, well.. you know what happens. Not often, but still, and that's not a VS Code issue, but seems to happen more with it than other apps because I'm selecting all or saving code. There are other keys in that area which are used quite a bit as well.

Also, I've tried looking the window events with BTT and trying to map a gets focus or moves into event with the panels, and sending a click event to VS Code, with no success, but maybe someone else may have better luck that I did.

There's is also the free option in Kariainer, but I never could get it to work correctly for me.

The other one which annoys me is you can't remap the Fn+C in macOS and it always brings up the Control Center, and I'd like it to do a Control+C instead, even BTT can't remap that, it's deep in the OS it seems.

> I have been following this thread for some time and the only explanation I can come up with is this con$piracy in some Micro$oft product development center where they do not _want_ to keep the UI flexible or adaptable for the user ... exactly on places where it _does_ hurt. For rea$on$ of dominating the market.

Well, VS Code is actually open source and free, so probably not the money. I've made this suggestion to the Cursor and Windsurf "suggestions" and people have upvoted it in those locations.  Since it's open source, money may not be a driver. However, it does go against their UI behavior guidelines for Windows I think. If someone with the Windsurf or Cursor teams could contribute this it would be great.  I use both of them and they are VS Code clones, and the developers there know the codebase fairly well I would assume.

I did actually clone the VS Code repo and had a few minutes to have a look, found references to various events and focus, but couldn't find an obvious place to start.  Not that familiar with TypeScript, so it will be put off for another day. All the bits seem to be there though, I'm just ignorant in the language and don't know the codebase well enough. Besides the pain point has not got to the point to make me dive in and spend a week or so on it, though it did get me to clone the repo and have a look about a week ago.

Maybe one day when I get some spare time I'll look into it more because I am constantly moving between panels and quite often out of habit move into the Terminal while the AI chat has focus and begin typing something like a `cd` or an `ls` and hit enter quickly because I'm looking for something to add to my AI chat prompt, so my prompt gets sent prematurely without all the information I wanted to send - and as a bonus, it will try to explain that it can help me with the `cd` or `ls` command if I'd like it to... *sigh* This is the biggest problem with the lack of focus follows mouse, mainly due to integrated AI along with the edit panels and terminal, constantly switching between them, and it breaks the workflow when I have it everywhere else.

Sorry for the length, but I have tried a few work-around on this and actually looked at the code a bit too.

**Comment by mwerle** [comment]
I literally can't fathom why this won't get merged/implemented. I understand some people may not like focus-follows-mouse which is why it's a no-brainer to make it an option instead of a change.

Maybe there's a good reason for not implementing this. If so, we haven't been provided with one.The only one I recall was something to do with the complexity and potential for impacting existing workflows or something along those lines. In the meantime we've had _far_ more disruptive and larger-scale changes added to VSCode, so that particular argument falls pretty flat.

I am yet another user who has, since discovering it, always had focus-follows-mouse enabled, in Windows as well as Unix/Linux. So the lack of support makes using VSCode quite painful at times and I type text into the wrong pane quite regularly, sometimes with devastating results as I also use vi-mode..

So please reconsider implementing this.

**Comment by SerhiiYukhymuk** [comment]
I would like to see this feature as same as many people above. I've been using "focus-follow-mouse" for quite some time now and it's a really good feature for providing a fast working environment in my system, so I've set it up wherever possible. It is very sad that this innovation is still waiting for approval.

**Comment by peter-lyons-kehl** [comment]
# Limited workaround
- only for window managers or OS'es that support focus-follows-mouse
- worth knowing even with other window managers or OS'es
- only for files in a workspace
- only for code/text editor (editing areas) and previews
- not for (VS Code built-in) terminal
- not for Explorer or other sidebars
- based on a new VS Code feature (probably since early 2025, or even only since mid 2025).

(All keyboard shortcuts here are from VS Code on Linux.)
1. Open your workspace/folder with menu "File > Open Folder..." (or Ctrl+K Ctrl+O), or (probably) with menu File > Open Workspace from File...".
   - Do NOT open files from outside the folder/workspace with menu "File > Open File..." (or Ctrl+O).
   - Do NOT open a new window with menu "File > New Window..." (or Ctrl+Shift+N) and
     -  do not then open files with menu "File > Open File..." (or Ctrl+O) in such windows. Such editors do not share an editing "buffer" with the main (project) window, so they will NOT show or share any unsaved changes.
     - do not even move an existing editor from the main project window (or from a window derived from it, see below) into such a new window, because then such an editor stops sharing unsaved changes with the main project window and its friends.
2. Have only one "group" (panel?) of "editors" (that is, one set of tabs) per window. Don't edit in split editors (so don't have multiple sets of tabs per window).
3. Whenever you want a new editor other than in the main window:
   1. Open the file in the main project window (where you opened the project folder/workspace and where explorer is).
   2. Move the editor into a new window
      - Right click at the tab > "Move into New Window.", or Ctrl+Shift+P > "View: Move Editor into New Window". Or,
      - If you already have an existing separate window, drag and drop the tab there. Or,
      - If you'd like to duplicate the editor, click the "Split Editor Right" button (right of the tabs), or Ctrl+Shift+P > "View: Split Editor". Then move the new editor to a separate window. Or,
      - You can split any editor in any window you've opened with the previous steps, and then drag and drop such a new editor to another/new window.

# VS Code version
Such editors share any unsaved changes and the undo/redo history. That seems to be a new/recent VS Code feature. It's present at least in version `1.102.2`.

# Some drawbacks
This does add a few extra steps per editor. But, if you want focus-follow-mouse (other than between editors and terminal/sidebar), you already are taking some steps to position your editors.

Depending on your window manager/OS, this also "wastes" some space on extra window titles/margins.

# GNOME
- [Tweaks (GNOME Tweaks)](https://gitlab.gnome.org/GNOME/gnome-tweaks) application and its menu Windows > Window Focus > "Focus on Hover" or "Focus Follows Mouse". (There also may be a related `dconf` setting.)
- [No Titlebar when Maximized](https://github.com/alecdotninja/no-titlebar-when-maximized) (which works even on VS Code windows that are not maximized, for example if you have one VS Code window positioned to the left with windows_key+left_arrow and another VS Code window positioned to the right with windows_key+right_arrow)
- various GNOME shell extensions that help with windows tiling
- `"window.customTitleBarVisibility": "never"` in `settings.json` or Ctrl+Shift+P > Preferences: Open User Settings > "Window: Custom Title Bar Visibility" set to "never" helps save screen space.

# Related tips for any/most window managers or OS'es
- VS Code has keyboard shortcuts to switch between editor groups (Ctrl+1, Ctrl+2...).
- See your window manager/OS shortcuts to switch between windows of the same application (in GNOME it's Alt+backtick).
- Toggle the VS Code built-in terminal with Ctrl+backtick.
- Toggle that panel (and whatever has been last selected in that panel, which may be the built-in terminal, error reports...) with Ctrl+J.

# Yes, limited
I do NOT to minimize the importance of, and demand for, native foccus-follow-mouse in VS Code. Those of us who can't use a window manager or OS that supports foccus-follow-mouse (or would be disadvantaged by it) have a legitimate need.

Feel free to correct my terminology or to mention any related/alternative tips.

**Comment by emogenet** [comment]
Has anyone tried to let an agentic AI loose on the VScode codebase asking it to actually implement this?
It's been 2 years an a half since I last tried to use VScode, it's as painful as ever, making VScode basically unusable for me.
I really can't believe the number of wasted clicks (and the wasted time) this issue has created over the years.

**Comment by rebroad** [comment]
9th birthday coming up! Maybe this isn't happening because it's Microsoft and they've always been click-to-focus?!

**Comment by stringcombiner** [comment]
Would definitely still love this. I came here after accidentally closing the primary file that I am editing, rather than the one I had opened to take a quick look at, for about the seventh time today. Like I do every day. Was really surprised to find that my answer to "how do I enable focus-follows-mouse in VScode?" is, apparently, "You can't." 

VScode's UI is largely excellent. The lack of a focus-follows-mouse option (especially without any credible explanation for why the feature won't be implemented) is really baffling. Any chance of any kind of update? @isidorn ?

**Comment by cyann** [comment]
Workaround I've been trying for the last few days: dim unfocused editors and panels.

I've added that in `settings.json`:

```json
{
	"accessibility.dimUnfocused.enabled": true,
	"accessibility.dimUnfocused.opacity": 0.60,
}
```

Play around with the opacity level, theme choice and light/dark mode may need different levels.

Sauce: https://github.com/microsoft/vscode-docs/blob/main/docs/configure/accessibility/accessibility.md#dim-unfocused-editors-and-terminals

**Comment by unixwzrd** [comment]
Focus-follows-mouse has existed in Unix/X11 environments for decades, like close to 50 years, so the ergonomics are well understood.
> Nearly another year down since my last comment, and still no actual update on this. Seriously would be using a different IDE by now (and privately am) but my dayjob is too invested in this one.

I feel your pain, and you rattled my cage... lol

What are you using instead?  Just curious, is it like VSCode? or do I have to learn a whole new IDE I've invested a great deal of time in.

> My workaround has been to use an external terminal which completely negates the advertised usefulness of the integrated terminal - and there are benefits which I definitely miss - but at least ensures I don't accidentally type editor commands into my terminal or vice-versa.
> 
> [@microsoft](https://github.com/microsoft): if there _is_ a good reason for not implementing this option (and please don't give us more waffle about bug surface or complexity as far more invasive changes have been made to the VSCode UI in the meantime), then could you _please_ share that with us?

Precisely, I keep iTerm2 at the side of my VSCode session and am using "Auto Raise" https://github.com/sbmpost/AutoRaise for years, and works fine, but yeah, kinda defeats the point. But the integrated terminal kinda sucks too TBH, though outside the scope of this thread. 😜

iTerm on the left and ChatGPT on the right, with VSCode in the middle and on top. I guess I use iTerm2 and ChatGPT to break "ties" in Cursor...

I haven't yet turned Codex loose on it, but may soon as the annoyance factor of this gets me every day. It's a terrible waste of my tokens, but might bring me fame (fortune doubtful) in the VSCode user community.

> Workaround I've been trying for the last few days: dim unfocused editors and panels.
> 
> I've added that in `settings.json`:
> 
> {
> 	"accessibility.dimUnfocused.enabled": true,
> 	"accessibility.dimUnfocused.opacity": 0.60,
> }
> Play around with the opacity level, theme choice and light/dark mode may need different levels.
> 
> Sauce: https://github.com/microsoft/vscode-docs/blob/main/docs/configure/accessibility/accessibility.md#dim-unfocused-editors-and-terminals

Nice "sauce" 😉 now it's not a secret anymore...

Well, that doesn't really eliminate the need to click to focus, but hey, whatever works I suppose. I'd prefer not to introduce more time delays to things or distraction. Nice idea though.

 But if they could implement that, why not this?

Better idea:

```json
 {
	"accessibility.focusFollowsMouse": true,
	"accessibility.raiseOnFocus": true,
}
```

But, sadly, that doesn't work until it gets implemented. I can dream though. Kinda makes you wonder.

> I don't see any existing PRs that would implement this feature. Does anyone know if there are any? If there aren't, do we know what work is needed to implement this?

There actually was a fix somewhere or in one of the many other focus threads. Someone even said they had a co-worker who implemented this, but the company wouldn't let them use it because it because it would be "non-standard". The whole UX is non-standard and allows for much configuration, and none of us probably likes the others' color choices too. But I would never force mine on anyone else. 😉

Sadly the individual with that disappeared and no PR was ever files, that I know of.

### My big question is why haven't the folks at Cursor or Windsurf taken a crack at this?
or one of the others which uses VSCode as a fork. They could probably get +43 subscribers (participants in this thread) by doing this maybe more. Well, 42 since I use Cursor now, though re-thinking that as I use Codex within Cursor and Cursor AI less and less. I could probably just use vanilla VSCode when I think about it.  (Hey cursor development teams, you listening?)

I actually like the ability of VSCode (and Cursor and Windsurf) to `ssh` into a remote host in order to do development and debugging, a very nice feature really.

**Comment by RoFz** [comment]
Hi @isidorn / @meganrogge, I know feature PRs for backlog `feature-request` issues need the issue owner's agreement before submission, so I'd like to ask before opening anything.

I've implemented a working proof-of-concept for focus-follows-mouse and would like to know (a) whether an external contribution for this would be considered, and (b) whether the design below matches what you'd want, before I invest further.

**Proposed design**

- New setting `accessibility.focusFollowsMouse` (boolean, default `false`, application scope, `accessibility` tag).
- A workbench contribution registered at the `Restored` lifecycle phase.
- When enabled, hovering a workbench part moves keyboard focus to it after a short debounce (about 200ms) so brief mouse transits don't steal focus:
  - Side Bar, Panel, and Auxiliary Bar via `layoutService.focusPart`
  - Editor groups via `group.focus()` (new and removed groups tracked dynamically)
- Guards: no focus shift while a mouse button is held (drag or selection), or while a context menu or overlay is open. A `mouseleave` cancels a still-pending editor-group focus.
- Reacts to the setting being toggled at runtime. Covered by unit tests and validated in a local dev build.

**Open questions for you**

1. Terminal is currently handled at the Panel level. Do you want finer-grained per-terminal behavior, given the issue specifically calls out terminals?
2. Is a single boolean the right granularity, or would you prefer per-area opt-in?
3. Should the debounce delay be configurable, and is about 200ms a sensible default?
4. Any concerns around auxiliary or floating windows?

Happy to sign the CLA and adapt the implementation to your guidance. Branch is ready to share if useful.

**Comment by guillaumeportails** [comment]
> I've implemented a working proof-of-concept for focus-follows-mouse

Funtastic, thanks @RoFz !

**Comment by unixwzrd** [comment]
@RoFz This is fantastic news, a bit back I tried making an extension, but as you point out, it's really dug in hard into the codebase. You're already a legend amongst the people wanting this, I will keep fingers crossed your PR is accepted!

**Comment by mwerle** [comment]
@RoFz , awesome work! We can only hope and pray that your PR is accepted.

1. What do you mean? In case there are multiple terminals open? In which case yes, whichever terminal has the mouse over it should be the active one..
2. All or nothing is fine by me.
3. Personally I'd like to see this configurable, but I'll take what I can get :)
4. ?

**Comment by meganrogge** [comment]
Hi @RoFz thanks for looking into this! We'd be open to an external PR for sure. 

Let's make debounce configurable and 200ms sounds good to me. I think just one setting for this makes sense. 

The initial implementation could skip per terminal handling for simplicity. But i'll leave that with you!

**Comment by RoFz** [comment]
@isidorn / @meganrogge I've opened the PR: https://github.com/microsoft/vscode/pull/317194

Thanks for the guidance. Here is how the open questions from my earlier comment were resolved:

1. Terminal granularity: deferred per your suggestion to skip per-terminal handling for simplicity. The terminal is still covered at the Panel level.
2. Single vs per-area setting: implemented as one setting, `accessibility.focusFollowsMouse.enabled`, as you preferred (no per-area split).
3. Configurable delay: added `accessibility.focusFollowsMouse.delay` (milliseconds, default 200, which you confirmed is sensible). It is read live, so changes apply without reload.
4. Auxiliary/floating windows: editor groups are covered across windows via IEditorGroupsService; the Side Bar/Panel/Auxiliary Bar listeners are main-window only.

## Known gaps and planned follow-ups

This PR is intended as a solid initial implementation rather than the complete picture. Items I would like to address in a future iteration, based on your feedback:

- Per-terminal focus handling (currently Panel-level only).
- Side Bar/Panel/Auxiliary Bar coverage in auxiliary/floating windows (currently main-window only; documented in code).
- Verification on Windows, macOS, web, and remote (manually verified on Linux so far; the code is platform-agnostic browser-layer logic, but I have not personally tested those targets).

Happy to keep iterating on these in subsequent changes once the core lands, and to adjust scope based on what you would prefer to see in the initial version.

The change has unit tests and I verified it manually with real mouse and keyboard input. I will sign the CLA when the bot prompts. Open to any review feedback.

**Comment by unixwzrd** [comment]
@RoFz

The terminal is kinda funky anyway, has quirks, but boy would it be nice, even for a single terminal it it sits in its own panel. I agree with @mwerle regarding terminal behavior, but will take the panel too, use iTerm2 anyway mostly. There are those times though when even a single terminal is good enough.

Main window is fine with me as long as the panel gets focus.

How can one get this nice shiny toy? I wanna test...

 macOS Tahoe 26.5, just updated the other day, I'll verify it works there. 😉

**Comment by unixwzrd** [comment]
@RoFz Any updates?
